diff --git a/android/app/build.gradle b/android/app/build.gradle
index a529cc9..80dcb66 100644
--- a/android/app/build.gradle
+++ b/android/app/build.gradle
@@ -96,6 +96,8 @@ android {
         versionName "1.1.0"
 
         buildConfigField "String", "REACT_NATIVE_RELEASE_LEVEL", "\"${findProperty('reactNativeReleaseLevel') ?: 'stable'}\""
+        buildConfigField "String", "GRAFANA_OTLP_API_KEY", "\"${GRAFANA_OTLP_API_KEY ?: ""}\""
+        buildConfigField "String", "GRAFANA_OTLP_ENDPOINT", "\"${GRAFANA_OTLP_ENDPOINT ?: ""}\""
     }
     signingConfigs {
         debug {
@@ -104,6 +106,14 @@ android {
             keyAlias 'androiddebugkey'
             keyPassword 'android'
         }
+        release {
+            if (project.hasProperty('MYAPP_UPLOAD_STORE_FILE')) {
+                storeFile file(MYAPP_UPLOAD_STORE_FILE)
+                storePassword MYAPP_UPLOAD_STORE_PASSWORD
+                keyAlias MYAPP_UPLOAD_KEY_ALIAS
+                keyPassword MYAPP_UPLOAD_KEY_PASSWORD
+            }
+        }
     }
     buildTypes {
         debug {
@@ -112,7 +122,7 @@ android {
         release {
             // Caution! In production, you need to generate your own keystore file.
             // see https://reactnative.dev/docs/signed-apk-android.
-            signingConfig signingConfigs.debug
+            signingConfig signingConfigs.release
             def enableShrinkResources = findProperty('android.enableShrinkResourcesInReleaseBuilds') ?: 'false'
             shrinkResources enableShrinkResources.toBoolean()
             minifyEnabled enableMinifyInReleaseBuilds
@@ -156,6 +166,9 @@ dependencies {
     // The version of react-native is set by the React Native Gradle Plugin
     implementation("com.facebook.react:react-android")
 
+    api(platform("io.opentelemetry.android:opentelemetry-android-bom:0.16.0-alpha"))
+    implementation("io.opentelemetry.android:android-agent") // Version is resolved thru bom
+
     def isGifEnabled = (findProperty('expo.gif.enabled') ?: "") == "true";
     def isWebpEnabled = (findProperty('expo.webp.enabled') ?: "") == "true";
     def isWebpAnimatedEnabled = (findProperty('expo.webp.animated') ?: "") == "true";

diff --git a/android/app/src/main/java/com/raskitech/cruisechat/MainApplication.kt b/android/app/src/main/java/com/raskitech/cruisechat/MainApplication.kt
index 6767f33..4924758 100644
--- a/android/app/src/main/java/com/raskitech/cruisechat/MainApplication.kt
+++ b/android/app/src/main/java/com/raskitech/cruisechat/MainApplication.kt
@@ -1,6 +1,7 @@
 package com.raskitech.cruisechat
 
 import android.app.Application
+import android.content.Context
 import android.content.res.Configuration
 
 import com.facebook.react.PackageList
@@ -16,41 +17,81 @@ import com.facebook.react.defaults.DefaultReactNativeHost
 import expo.modules.ApplicationLifecycleDispatcher
 import expo.modules.ReactNativeHostWrapper
 
+import java.time.Duration
+import java.util.concurrent.TimeUnit
+import android.util.Log
+
+import io.opentelemetry.android.agent.OpenTelemetryRumInitializer
+import io.opentelemetry.android.OpenTelemetryRum
+import io.opentelemetry.api.common.Attributes
+import io.opentelemetry.api.common.AttributeKey.stringKey
+import io.opentelemetry.api.GlobalOpenTelemetry
+
+
 class MainApplication : Application(), ReactApplication {
 
-  override val reactNativeHost: ReactNativeHost = ReactNativeHostWrapper(
-      this,
-      object : DefaultReactNativeHost(this) {
-        override fun getPackages(): List<ReactPackage> =
-            PackageList(this).packages.apply {
-              // Packages that cannot be autolinked yet can be added manually here, for example:
-              // add(MyReactNativePackage())
-            }
+    override val reactNativeHost: ReactNativeHost = ReactNativeHostWrapper(
+        this,
+        object : DefaultReactNativeHost(this) {
+            override fun getPackages(): List<ReactPackage> =
+                PackageList(this).packages.apply {
+                    // Packages that cannot be autolinked yet can be added manually here, for example:
+                    // add(MyReactNativePackage())
+                }
+
+            override fun getJSMainModuleName(): String = ".expo/.virtual-metro-entry"
 
-          override fun getJSMainModuleName(): String = ".expo/.virtual-metro-entry"
+            override fun getUseDeveloperSupport(): Boolean = BuildConfig.DEBUG
 
-          override fun getUseDeveloperSupport(): Boolean = BuildConfig.DEBUG
+            override val isNewArchEnabled: Boolean = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
+        }
+    )
 
-          override val isNewArchEnabled: Boolean = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED
-      }
-  )
+    override val reactHost: ReactHost
+        get() = ReactNativeHostWrapper.createReactHost(applicationContext, reactNativeHost)
 
-  override val reactHost: ReactHost
-    get() = ReactNativeHostWrapper.createReactHost(applicationContext, reactNativeHost)
+    override fun onCreate() {
+        super.onCreate()
+
+        val endpoint = BuildConfig.GRAFANA_OTLP_ENDPOINT
+        val apiKey = BuildConfig.GRAFANA_OTLP_API_KEY
+        Log.d("MainApplication", "Initializing OpenTelemetry RUM with endpoint: $endpoint")
+
+        val rum = OpenTelemetryRumInitializer.initialize(
+            context = this@MainApplication,
+            configuration = {
+                httpExport {
+                    baseUrl = endpoint
+                    baseHeaders = mapOf(
+                        "Authorization" to "Bearer $apiKey"
+                    )
+                }
+                instrumentations {
+                    activity {
+                        enabled(false)
+                    }
+                    fragment {
+                        enabled(false)
+                    }
+                    slowRenderingReporter {
+                        enabled(false)
+                    }
+                }
+            }
+        )
+        GlobalOpenTelemetry.set(rum.getOpenTelemetry())
+
+        DefaultNewArchitectureEntryPoint.releaseLevel = try {
+            ReleaseLevel.valueOf(BuildConfig.REACT_NATIVE_RELEASE_LEVEL.uppercase())
+        } catch (e: IllegalArgumentException) {
+            ReleaseLevel.STABLE
+        }
+        loadReactNative(this)
+        ApplicationLifecycleDispatcher.onApplicationCreate(this)
+    }
 
-  override fun onCreate() {
-    super.onCreate()
-    DefaultNewArchitectureEntryPoint.releaseLevel = try {
-      ReleaseLevel.valueOf(BuildConfig.REACT_NATIVE_RELEASE_LEVEL.uppercase())
-    } catch (e: IllegalArgumentException) {
-      ReleaseLevel.STABLE
+    override fun onConfigurationChanged(newConfig: Configuration) {
+        super.onConfigurationChanged(newConfig)
+        ApplicationLifecycleDispatcher.onConfigurationChanged(this, newConfig)
     }
-    loadReactNative(this)
-    ApplicationLifecycleDispatcher.onApplicationCreate(this)
-  }
-
-  override fun onConfigurationChanged(newConfig: Configuration) {
-    super.onConfigurationChanged(newConfig)
-    ApplicationLifecycleDispatcher.onConfigurationChanged(this, newConfig)
-  }
 }
